scipy/stats/_hypotests.py

```diff
@@ -663,3 +664,318 @@ def somersd(x, y=None):
+    elif alternative == "greater":
+        index_arr = wald_statistic >= wald_stat_obs
+    else:
+        msg = (
+            "`alternative` should be one of {'two-sided', 'less', 'greater'},"
+            f" found {alternative!r}"
+        )
+        raise ValueError(msg)
+
+    p_value = _binomial_maximisation_of_p_value_with_nuisance_param(
+        total_col_1, total_col_2, index_arr, n_iter
+    )
+    return BarnardExactResult(wald_stat_obs, p_value)
+
+
+def _binomial_maximisation_of_p_value_with_nuisance_param(
+    total_c1, total_c2, idx, n_iter
+):
+    r"""
+    Maximisation of the pvalue in respect of a nuisance parameter considering
+    a 2x2 sample space.
+
+    Barnard exact test iterate over a nuisance parameter
+    :math:`\pi \in [0, 1]` to find the maximum p-value. To search this
+    maxima, I am using a grid search algorithm, reducing :math:`\pi`
+    bounds after each
+    iterations `n_iter`. To compute the p-value, I am using numpy ndarrays
+    which offer powerful broadcast operations. The different nuisance
+    parameters are stored in an ndarray of length 100 and of shape (1, 1,
+    100). Also, to compute the different combination used in the
+    p-values' computation formula, uses `gammaln` which is
+    more tolerant for large value than `scipy.special.comb`. This gives
+    a log combination. For the little precision lost, we gain a lot of
+    performance.
+    """
+    n = total_c1 + total_c2
+    x1 = np.arange(total_c1 + 1, dtype=np.int64).reshape(-1, 1, 1)
+    x2 = np.arange(total_c2 + 1, dtype=np.int64).reshape(1, -1, 1)
+
+    x1_log_comb = _compute_log_combinations(total_c1)
+    x2_log_comb = _compute_log_combinations(total_c2)
+    x1_sum_x2_log_comb = x1_log_comb[x1] + x2_log_comb[x2]
+
+    def get_pvalue_from(nuisance_param):
+
+        with np.errstate(divide="ignore", invalid="ignore"):
+            log_nuisance_arr = np.log(
+                nuisance_param,
+                out=np.zeros_like(nuisance_param),
+                where=nuisance_param >= 0,
+            )
+            log_1_minus_nuisance_arr = np.log(
+                1 - nuisance_param,
+                out=np.zeros_like(nuisance_param),
+                where=1 - nuisance_param >= 0,
+            )
+
+            nuisance_power_x1_x2 = log_nuisance_arr * (x1 + x2)
+            nuisance_power_x1_x2[(x1 + x2 == 0)[:, :, 0]] = 0
+
+            nuisance_power_n_minus_x1_x2 = log_1_minus_nuisance_arr * (
+                n - x1 - x2
+            )
+            nuisance_power_n_minus_x1_x2[(x1 + x2 == n)[:, :, 0]] = 0
+
+            tmp_values_arr = np.exp(
+                x1_sum_x2_log_comb
+                + nuisance_power_x1_x2
+                + nuisance_power_n_minus_x1_x2
+            )
+
+        tmp_values_arr /= tmp_values_arr.sum(axis=(0, 1)).reshape(1, 1, -1)
+        # This operation compensate numerical errors because sums of
+        # p_values_arr should always be equal to one.
+
+        # Just sum where TX >= TX0
```

Comment:
We don't use these names anymore.

Reply:
Remove this line as it was more confusing than helpful